### PR TITLE
Allow using visible regions with projections

### LINKED ISSUE #3073: Allow external visible regions in ProjectionViewer
- [X] I verified ~~[latest Integration Build of Eclipse SDK](https://download.eclipse.org/eclipse/downloads/)~~ current `master` doesn't provide the feature.

## Suggestion

The `ProjectionViewer` allows adding folding regions (projections) but it also allows setting the visible regions using the `setVisibleRegion()` method. However, when `setVisibleRegion` is called, [projections are disabled automatically](https://github.com/eclipse-platform/eclipse.platform.ui/blob/a4840da1b2b89a93298ea766feb7f2738c935677/bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java#L687), presumably because `ProjectionViewer` needs to hide parts of the document when folding regions are collapsed.

To see what happens when visible regions are used together with folding, consider the following test (added as an additional method to [`ProjectionViewerTest`](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/tests/org.eclipse.jface.text.tests/src/org/eclipse/jface/text/tests/ProjectionViewerTest.java)):
```java
	@Test
	public void testProjectionWithVisibleRegion() {
		Shell shell= new Shell();
		shell.setLayout(new FillLayout());
		ProjectionViewer viewer= new ProjectionViewer(shell, null, null, false, SWT.NONE);
		String documentContent= """
				Hello
				World
				123
				456
				""";
		Document document= new Document(documentContent);
		viewer.setDocument(document, new AnnotationModel());
		int regionLength= documentContent.indexOf('\n');
		viewer.setVisibleRegion(0, regionLength);
		viewer.enableProjection();
		viewer.getProjectionAnnotationModel().addAnnotation(new ProjectionAnnotation(false), new ProjectionPosition(document));
		shell.setVisible(true);
		// viewer.getTextOperationTarget().doOperation(ProjectionViewer.COLLAPSE_ALL); // uncommenting this leads to a different test failure
		viewer.getTextOperationTarget().doOperation(ITextOperationTarget.SELECT_ALL);
		try {
			assertEquals(0, viewer.getVisibleRegion().getOffset());//If collapsed: expected 0 but was 6
			assertEquals(regionLength, viewer.getVisibleRegion().getLength());//expected 5 but was 20
		} finally {
			shell.dispose();
		}
	}
```
This test enables folding (`enableProjection()`) _after_ calling `setVisibleRegion()` (this being allowed may be a bug) and shows what happens when using both projections and visible regions together. (The Javadoc of `ProjectionViewer` does not seem to suggest that folding with a custom visible region is not supported so I guess this could also be seen as a bug.)

I suggest that the public `setVisibleRegion` is changed in a way such that:
- It hides everything before and after the passed region
- Folding regions can be collapsed as it is possible without `setVisibleRegion`.

It might be possible to internally implement this by adding an additional collapsed region from the document to the start of the given visible region and one from the end of the visible region to the end of the visible document.

This is motivated by https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2264 as JDT uses both folding and also has a "Only show selected Java element" that calls `setVisibleRegion` to ensure only the selected element is shown. If this feature is enabled, [folding is generally disabled if "Only show the selected element" is enabled](https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/fc9808007c93d80d8e56dc421175f528296d418a/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaEditor.java#L1878).

To test it, the `(store == null || !store.getBoolean(PreferenceConstants.EDITOR_SHOW_SEGMENTS))` check could be removed from https://github.com/eclipse-jdt/eclipse.jdt.ui/blob/fc9808007c93d80d8e56dc421175f528296d418a/org.eclipse.jdt.ui/ui/org/eclipse/jdt/internal/ui/javaeditor/JavaEditor.java#L1878. After that, folding _should_ (in theory) work with "Only show the selected Java element" (and selecting different parts of while the document is open).

## Community

- [X] I understand suggesting an enhancement doesn't mandate anyone to implement it. Other contributors may consider this suggestion, or not, at their own convenience. The most efficient way to get it fixed is that I implement it myself and contribute it back as a good quality patch to the project.

### PR DESCRIPTION
**Pull request by danthe1st**
This is my attempt to fix #3073 in order to implement https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2264

### Description

`ProjectionViewer` currently cannot use visible regions and projections together (except by calling `enableProjections()` after `setVisibleRegions` in which case it will show a wrong region). This PR changes the implementation of visible regions to collapse everything outside the visible region.

### Note

This PR does not allow collapsing or expanding any projection/folding regions outside of the declared _visible region_ (e.g. by other code modifying the document in some way or if a projection region is partially in the visible region).

Aside from that, this implementation has the following effect: If `setVisibleRegion` is called with a region that is fully collapsed, the editor is basically empty. If another behavior is desired, it would be good to know what the better behavior is.

### testing with JDT

To test this with JDT as requested in https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2264, do the following:
- Create a workspace with eclipse.platform.ui and JDT-UI set up (I did so by creating a Target Platform and added whatever I thought was necessary until I no longer got any errors but I have no idea how to properly do that)
- Use the changes in eclipse.platform.ui
- Apply https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302 with JDT-UI:
- Create an "Eclipse Application" run configuration with this version of Eclipse Platform and the patched version of JDT-UI and start it
- Enable Window > Preferences > Java > Editor > "Only show the selected Java Element"
- Make sure folding is enabled at Window > Preferences > Java > Editor > Folding
- Open a Java class
- Select elements in the outline and observe that only the "selected element" is shown in the editor and folding still works (whether it's normal folding, custom folding regions, "Extended Folding", etc shouldn't make a difference)
image
- If "Enhanced Folding" is active, make sure that https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2410 is available as well

### CODE REVIEW

**Comment by iloveeclipse** [comment]
@danthe1st : thanks for PR, few things:

1. Please rebase on master
2. Please squash two commits into one
3. Please update commit message to follow commit message guidelines: https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

**Comment by danthe1st** [comment]
> @danthe1st : thanks for PR, few things:
> 
>     1. Please rebase on master
> 
>     2. Please squash two commits into one
> 
>     3. Please update commit message to follow commit message guidelines: https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#commit-message-recommendations

I'll make sure to do that once I located and fixed a remaining issue in combination with JDT (if I select a method that's within a custom folding region and enter text, it seems to show "too much" text for some reason). While I think that's an issue with JDT-UI (it only seems to happen with custom folding regions _and_ extended folding), I want to be sure about it first (and I opened this PR before everything is finished to make it visible in advance/allow for discussion if applicable).

**Comment by danthe1st** [comment]
I have made the requested changes. However, I want to note the following:
- This has an impact on behavior(al compatibility). If some project/plugin/whatever calls `enableProjections()` and `setVisibleRegion` on some `ProjectionViewer`, this PR affects what is shown to the user.
- I didn't do performance testing.
- I think it would be a good idea for someone from JDT (or other projects if make use of this?) to test it before this is merged.

**Comment by danthe1st** [comment]
I am not sure why the Jenkins build seems to have issues getting tycho set up (maybe issues with https://repo.eclipse.org/?)

**Comment by merks** [comment]
Yes there are infrastructure problems again:

- https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/6377
- https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/6378

I restarted the build.

**Comment by merks** [comment]
It looks like there was a test failure:

https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-3074/lastCompletedBuild/testReport/

**Comment by danthe1st** [comment]
I now switched to an approach that should have fewer side effects. Instead of adding additional projection regions, I am now using the `expand()` and `collapse()` methods directly.

**Comment by danthe1st** [comment]
@merks Would you or someone else care to review this?

**Comment by merks** [comment]
> @merks Would you or someone else care to review this?

I don’t really feel qualified to judge this. ☹️ I’m not sure whom to suggest.

**Comment by danthe1st** [comment]
> I’m not sure whom to suggest

Yeah I have no idea either and I don't want to randomly start ping people out of nowhere (so I tried to ask people who already commented i.e. you). One option might be to look for people who contributed in that area at some point (doesn't really seem like this is the case for some more than for others?) or I could ask in the `platform-dev` mailing list.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 1]
```diff
@@ -683,9 +732,48 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
 
 	@Override
 	public void setVisibleRegion(int start, int length) {
-		fWasProjectionEnabled= isProjectionMode();
-		disableProjection();
-		super.setVisibleRegion(start, length);
+		if (isProjectionMode()) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: May I ask you to revert the logic and check/handle the other branch first with a quick return, so the followed code can be read without ome extra indentation  level?

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 2]
```diff
@@ -301,6 +334,8 @@ private void computeExpectedExecutionCosts() {
 	 */
 	private int fDeletedLines;
 
+	private UpdateDocumentListener fUpdateDocumentListener= new UpdateDocumentListener();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: can this be initialized in constructor please?

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 3]
```diff
@@ -510,6 +545,11 @@ public final void disableProjection() {
 			fProjectionAnnotationModel.removeAllAnnotations();
 			fFindReplaceDocumentAdapter= null;
 			fireProjectionDisabled();
+			if (fVisibleRegionDuringProjection != null) {
+				super.setVisibleRegion(fVisibleRegionDuringProjection.getOffset(), fVisibleRegionDuringProjection.getLength());
+				fVisibleRegionDuringProjection= null;
+			}
+			getDocument().removeDocumentListener(fUpdateDocumentListener);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: I see in other parts of code the getDocument() return value is checked for null. Your patch doesn't (here and below). Why?

**Comment by iloveeclipse** [comment]
> I now switched to an approach that should have fewer side effects. Instead of adding additional projection regions, I am now using the `expand()` and `collapse()` methods directly.

Could you please update PR description to match proposed implementation?

**Comment by danthe1st** [comment]
Thank you for your review. I made the requested changes and updated the description.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 4]
```diff
@@ -683,9 +740,52 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+				end++;
+			}
+
+			int endInvisibleRegionLength= documentLength - end;
+			if (endInvisibleRegionLength > 0) {
+				collapse(end, endInvisibleRegionLength, true);
+			}
+			fVisibleRegionDuringProjection= new Region(start, end - start);
+		} catch (BadLocationException e) {
+			e.printStackTrace();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: Please log exception properly, don't print to std error.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 5]
```diff
@@ -683,9 +740,52 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
-		super.setVisibleRegion(start, length);
+		if (!isProjectionMode()) {
+			super.setVisibleRegion(start, length);
+			return;
+		}
+		IDocument document= getDocument();
+		if (document == null) {
+			return;
+		}
+		try {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java:759]: Please add comment before the block about why we are doing that. That would make it easier to understand the following new chunk of code.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 6]
```diff
@@ -272,6 +272,34 @@ private void computeExpectedExecutionCosts() {
...
 
+	/**
+	 * An {@link IDocumentListener} that makes sure that {@link #fVisibleRegionDuringProjection} is
+	 * updated when the document changes and ensures that the collapsed region after the visible
+	 * region is recreated appropriately.
+	 */
+	private final class UpdateDocumentListener implements IDocumentListener {
+		@Override
+		public void documentChanged(DocumentEvent event) {
+			if (fVisibleRegionDuringProjection != null) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: Please also here invert the logic to have fast return and no extra indent block after that. Every indented level makes the code harder to read.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 7]
```diff
@@ -272,6 +272,34 @@ private void computeExpectedExecutionCosts() {
...
+				int oldLength= event.getLength();
+				int newLength= event.getText().length();
+				int oldVisibleRegionEnd= fVisibleRegionDuringProjection.getOffset() + fVisibleRegionDuringProjection.getLength();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: I haven't tried in editor yet, just curious what happens if the entire element is deleted, either in text or outside (like "delete method" in the outline view).

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java (hunk 5 above)
Comment (iloveeclipse) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java:759]: I wonder if the following block could be refactored to something like number of steps that explain the logic:

collapseBefore(region)
collapseAfter(region)
whateverElse()

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 8]
```diff
@@ -683,9 +740,52 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+				collapse(end, endInvisibleRegionLength, true);
+			}
+			fVisibleRegionDuringProjection= new Region(start, end - start);
+		} catch (BadLocationException e) {
+			e.printStackTrace();
+		}
+		fVisibleRegionDuringProjection= new Region(start, length);
+	}
+
+	private boolean isWhitespaceButNotNewline(char c) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: this is static

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 9]
```diff
@@ -683,9 +740,52 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+		return Character.isWhitespace(c) && !isLineBreak(c);
+	}
+
+	private boolean isLineBreak(char c) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: static too

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java (hunk 4 above)
Comment (danthe1st) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: I used `ILog` as I found something similar in another place.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java (hunk 7 above)
Comment (danthe1st) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: If it is deleted in the editor, it results in the element being removed from the editor (and the editor being empty) which is the same as before (JDT doesn't call `setVisibleRegion()` again in this case) without any issues.
If it is deleted outside, it works the same _most_ of the time but there is one edge case (if it is at the end of the file and the method is "big enough") where you can select the surrounding element and the closing `}` is missing. I'll take a look at that.

Comment (danthe1st) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: That edge case was an off-by-1 error that should now be fixed by changing `event.getOffset() + oldLength < oldVisibleRegionEnd` condition to `event.getOffset() + oldLength <= oldVisibleRegionEnd`)
I also added a test case for this.

**Comment by iloveeclipse** [comment]
Just a hint: for all the comments you have resolved, please use "Resolve Conversation" button so it is easier to see what is left open.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 10]
```diff
@@ -518,9 +566,18 @@ public final void disableProjection() {
 	 */
 	public final void enableProjection() {
 		if (!isProjectionMode()) {
+			IRegion visibleRegion= getVisibleRegion();
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: this line is not needed here but right before the `if (visibleRegion` block, so please move there.

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 11]
```diff
@@ -683,9 +744,73 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+		}
+		try {
+			// If the visible region changes, make sure collapsed regions outside of the old visible regions are expanded
+			// and collapse everything outside the new visible region
+			int end= computeEndOfVisibleRegion(start, length, document);
+			expandOutsideCurrentVisibleRegion(document);
+			collapseOutsideOfNewVisibleRegion(start, end, document);
+			fVisibleRegionDuringProjection= new Region(start, end - start);
+		} catch (BadLocationException e) {
+			String PLUGIN_ID= "org.eclipse.jface.text"; //$NON-NLS-1$
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: Please change to `ILog log= ILog.of(getClass());`

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 12]
```diff
@@ -683,9 +744,73 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+			Bundle bundle= Platform.getBundle(PLUGIN_ID);
+			ILog log= ILog.of(bundle);
+			log.log(new Status(IStatus.WARNING, getClass(), IStatus.OK, e.getMessage(), e));
+		}
+		fVisibleRegionDuringProjection= new Region(start, length);
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: is this line meant to be inside the catch block? Or why it overrides the value set in the try block?

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java [hunk 13]
```diff
@@ -683,9 +744,73 @@ private int toLineStart(IDocument document, int offset, boolean testLastLine) th
...
+		}
+	}
+
+	private static int computeEndOfVisibleRegion(int start, int length, IDocument document) throws BadLocationException {
+		int documentLength= document.getLength();
+		int end= start + length + 1;
+		// ensure that trailing whitespace is included
+		// In this case, the line break needs to be included as well
+		boolean visibleRegionEndsWithTrailingWhitespace= end < documentLength && isWhitespaceButNotNewline(document.getChar(end - 1));
+		while (end < documentLength && isWhitespaceButNotNewline(document.getChar(end))) {
```
Comment (iloveeclipse) [inline bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java:804]: this iterates over every character in the document, but shouldn't iterate by lines and stop immediately current line checks if the character is not matching?

Comment (danthe1st) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java:804]: It isn't (supposed to be) iterating over every character in the document but only the characters in that specific line (to find a non-whitespace character or the line break).
The goal is that _if_ the line ends with trailing whitespace, the region to collapse can only start at the next line (but still keep non-whitespace characters in the same line as it behaves correctly otherwise.
You can test it by creating a a visible region such that it has a nonzero amount of whitespace before a line break after the region ends (e.g. `ProjectionViewerTest#testVisibleRegionEndsWithWhitespace`).

Alternative code to test with JDT (note the tab character after the `}`):
```java
void test(){

}	
```

File: bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java (hunk 12 above)
Comment (danthe1st) [reply bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java]: That line shouldn't be there at all - removed.

**Comment by danthe1st** [comment]
I am not sure in what ways my changes are related (assuming they are related) to the test failure on MacOS (and I don't have a MacOS device to test it on).

It might be related to https://github.com/eclipse-platform/eclipse.platform.ui/issues/294

**Comment by totomomo18** [comment]
Any chance this feature be available in the next Eclipse version 2025-09 ?

**Comment by danthe1st** [comment]
> Any chance this feature be available in the next Eclipse version 2025-09 ?

I don't think so (and I don't think it would be a good idea even if that was reviewed and deemed ok content-wise now) since we are [approaching RC1](https://github.com/eclipse-simrel/.github/blob/main/wiki/SimRel/2025-09.md) and as this change could impact other parts of the IDE and maybe also RCP application, it would be ideal if it was merged before even M1 (so if there are issues originating from it, there is enough time to fix them before a release).

That being said, I think the implementation is at a point where it can be re-reviewed, especially since [this bugfix](https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2410) has been merged. One thing I'm not quite sure about (yet) is the following note in the PR description:
> Expanding a folding region outside the currently visible region will make that folding region visible. If wanted, I can change that.

I (think I) could make it so that expanding folding regions that may have parts outside of the visible regions just ignores these parts.

**Comment by totomomo18** [comment]
Thank you for your reply and all the great work you have done so far
So I guess there are still problem  with this. I just thought since it passed all the tests it is ready to be merged.

You said 
"Expanding a folding region outside the currently visible region will make that folding region visible. If wanted, I can change that."

But how is that possible since when you invoke show only selected element you only see the code of the selected element  on the file you are coding lets call the file file a. how can you expand on something outside of selected element on  file a ? unless you are saying that if you expand a region on a different file lets call it file b then it will disable the show only selected element on file a ?
That is really a no good file b should affect file  a. It should be fixed changed

Do you think there is any chance this feature will be merged in the next version of Eclipse after 202509 like 202512 ?
I hope it will cause I am looking forward for this feature.

**Comment by danthe1st** [comment]
It's possible this hasn't been merged until now because I didn't care enough to ask for a re-review :joy:  (I guess I should rebase and do that after the 2025-09 release)

> But how is that possible since when you invoke show only selected element you only see the code of the selected element on the file you are coding lets call the file file a. how can you expand on something outside of selected element on file a ? unless you are saying that if you expand a region on a different file lets call it file b then it will disable the show only selected element on file a ?

This PR is about the change in platform.ui which provides the APIs relevant for that. If a client decides they want to show projection regions in a way that these are not entirely within or outside of folding regions, they can do so (and in fact, this situation can also occur with custom folding regions where one comment is inside the visible region and the other isn't). This decision is something I'd like to get the opinions of the committers/possibly other people.

Regarding the tests, I think there are still 3 that didn't run with the current version yet and IIRC some tests failed there with MacOS on previous versions (and I don't have a MacOS device to fix these issues if they originate from this PR). These tests would likely be executed when a committer reviews it (or clicks on the button to approve the test run which has to be redone whenever a new version is pushed because I haven't contributed to platform.ui before).

Also I don't know how big the effects of this PR are on other code based on Eclipse that isn't JDT (e.g. companies building RCP applications).

> Do you think there is any chance this feature will be merged in the next version of Eclipse after 202509 like 202512 ?

¯\\\_(ツ)_/¯

**Comment by danthe1st** [comment]
@iloveeclipse Would you like to re-review this (and https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302) at some point when you have time? I think I forgot to mention that I have made the requested changes.

**Comment by danthe1st** [comment]
As I didn't get any feedback until now, I made the following choice (can still be changed if there are good reasons for it):

Since it is possible for changes outside of the visible region to happen (e.g. using custom actions, other code adding projection regions or anything else that might modify the document) and users wouldn't be interested to see stuff outside of the visible region, I decided to change the logic to ignore collapsing and expanding outside of the visible region.

I also adjusted the description accordingly.

> This PR does not allow collapsing or expanding any projection/folding regions outside of the declared _visible region_ (e.g. by other code modifying the document in some way or if a projection region is partially in the visible region).

**Comment by laeubi** [comment]
@mickaelistria can you help reviewing this?

**Comment by danthe1st** [comment]
It seems like this PR missed M1 now. Would it be possible to get this reviewed before M2 (with sufficient time left for https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302)?
@mickaelistria 

Or would it be better to wait for another release cycle (2026-03)? But even in that case, it would be better to get this reviewed at some point.

**Comment by mickaelistria** [comment]
I won't review in details but this patch adds many good tests, is actually a fix, has no new API and all payload changes are internal and isolated; and code is understandable enough. So let's merge and worth case, we can revert it.

**Comment by iloveeclipse** [comment]
This PR caused regression: https://github.com/eclipse-platform/eclipse.platform.ui/issues/3380
@mickaelistria, @danthe1st : could you please check?

**Comment by iloveeclipse** [comment]
This PR caused another regression: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/2532

@mickaelistria, @danthe1st : please follow up. The way the listener is added & removed is not OK, as it will be not **always** removed. A simple solution is doable, I've posted it on the bug, but I'm not sure if that is enough.

**Comment by mickaelistria** [comment]
@danthe1st @iloveeclipse I think we should give about a week for a fix to be pushed, and if nothing has happened before that, we can revert.

**Comment by cdietrich** [comment]
causes breaking change in Xtext comment editing strategy https://github.com/eclipse-xtext/xtext/issues/3531

**Comment by totomomo18** [comment]
is this feature included in the new Eclipse 2025 12 ? where is it  going to be included in a official release ?

**Comment by merks** [comment]
> is this feature included in the new Eclipse 2025 12 ? where is it going to be included in a official release ?

It's merged so I expect it is in the release, which will go live later today.

**Comment by totomomo18** [comment]
I can not wait :) Thanks for the replay

**Comment by danthe1st** [comment]
> is this feature included in the new Eclipse 2025 12 ? where is it going to be included in a official release ?

This feature s included in 2025-12 but the JDT one as some issues are still open. See https://github.com/eclipse-platform/eclipse.platform.ui/pull/3456 (my plan was to rebase that and test everything and some point this month/after the release but I didn't get to it) (and technically also https://github.com/eclipse-platform/eclipse.platform.ui/pull/3457#discussion_r2483750797 but I don't consider that to be a big issue).

if you really want to try it out, feel free to build JDT-UI from source (if you need help with that, you can contact me directly if you want to).

**Comment by totomomo18** [comment]
Just downloaded the new Sts and I tested it it seem to be working Thank You very much :)

Why don't they include this feature in the new features I looked for it in the New & Noteworthy page
https://eclipseide.org/release/noteworthy/

For me it is an important fix which is so useful allow to use both the folding and the show only selected element together at the same time. It really help you to focus on the code you want to work at the a specific method

**Comment by danthe1st** [comment]
> Just downloaded the new Sts and I tested it it seem to be working

It shouldn't do anything if you don't have the JDT-UI patch (or use a different editor making use of that) which isn't merged yet..? This is why there's no N&N about it in 2025-12.

This PR is just about giving Eclipse Platform the necessary capabilities for this. Of course, a custom editor may use that functionality.

But yes, if you apply the JDT-UI patch yourself and somehow install it into Spring Tools 5 (or Eclipse 2025-12), it should work.

**Comment by totomomo18** [comment]
It seem to be working but the closing  bracket of the method that is show only selected element seem to disappear

Before show selected element

image

After show selected element
image

**Comment by danthe1st** [comment]
> It seem to be working but the closing bracket of the method that is show only selected element seem to disappear

Oh, I think I understand why it "works" for you. This is just because you enabled the option while the file is opened. The JDT-UI change (which is not merged yet) makes sure that this feature works on newly opened files (basically, it removes some code that disables that feature). Until https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302 is merged, you cannot expect this to work. If you close and reopen the file, it shouldn't show any folding regions.
That being said, I don't know why exactly the `}` is hidden but I couldn't yet test it with the new version.

If a `}` is hidden, I _might_ be able to check it if I have the exact code (with the exact whitespace used, including on lines containing only whitespace).

But for Java-specific things, this (platform-UI) is (probably) not the right repository.

<sub>Note: I am just an outside contributor, I am not a maintainer of any Eclipse projects. I am just occasionally contributing to Eclipse in my free time (when I consider something to be important or interesting enough).</sub>

**Comment by totomomo18** [comment]
Should I report the disappearing bracket when  invoking show only selected element as bug to Eclipse or Sts Repository ? 
if yes which repository   ? 
I know that this is a new bug since disappearing bracket did not disappear in previous eclipse/sts version 

Also when do you think the patch of  JDT-UI patch  be merged ?  
Do you think that this patch will slove the  disappearing bracket  problem ?

**Comment by danthe1st** [comment]
> Should I report the disappearing bracket when invoking show only selected element as bug to Eclipse or Sts Repository ?
if yes which repository ?

If you can reproduce it with Eclipse 2025-12 and https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302 (you have to compile JDT-UI yourself), please add a comment in that issue (and hopefully, I can take a look). If it cannot be reproduced like that, I don't think it would be a bug as it isn't enabled anyways. Spring Tools would be the wrong repository.

> Also when do you think the patch of JDT-UI patch be merged ?

Whenever it's finished. It depends on my motivation. There's also another issue with it which would be good to resolve first.

Creating a new issue for something JDT-UI specific doesn't make sense as the relevant PR isn't merged yet (again, it only affects open editors when you enable the feature in the preferences).

**Comment by totomomo18** [comment]
> > Should I report the disappearing bracket when invoking show only selected element as bug to Eclipse or Sts Repository ?
> > if yes which repository ?
> 
> If you can reproduce it with Eclipse 2025-12 and [eclipse-jdt/eclipse.jdt.ui#2302](https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/2302) (you have to compile JDT-UI yourself), please add a comment in that issue (and hopefully, I can take a look). If it cannot be reproduced like that, I don't think it would be a bug as it isn't enabled anyways. Spring Tools would be the wrong repository.
> 
> > Also when do you think the patch of JDT-UI patch be merged ?
> 
> Whenever it's finished. It depends on my motivation. There's also another issue with it which would be good to resolve first.
> 
> Creating a new issue for something JDT-UI specific doesn't make sense as the relevant PR isn't merged yet.

I just downloaded Eclipse 2015 12 ( not Sts basic Eclipse ide in zip format) from the website. A clean install without your patch (because I dont know how to install it or compile and ide with a patch) and did the same test the folding sort of work with show only selected element and bracket still disappear it seem to be a bug that did happen before should I report it  as a bug ? and to which repository ?

Before invoking show only selected element on the main method
image

After invoking show only selected element on the main method
image

**Comment by danthe1st** [comment]
> A clean install without your patch (because I dont know how to install it or compile and ide with a patch) and did the same test the folding sort of work with show only selected element and bracket still disappear it seem to be a bug that did happen before should I report it as a bug ? and to which repository ?

Does it also happen if you close and reopen the editors _after_ setting the preference (without changing the preference again)? Also, it would be useful to know the exact code you are using (with whitespace) and the folding configuration. If not, I wouldn't consider this to be a fully implemented feature.

**Comment by totomomo18** [comment]
> > A clean install without your patch (because I dont know how to install it or compile and ide with a patch) and did the same test the folding sort of work with show only selected element and bracket still disappear it seem to be a bug that did happen before should I report it as a bug ? and to which repository ?
> 
> Does it also happen if you close and reopen the editors _after_ setting the preference? Also, it would be useful to know the exact code you are using (with whitespace) and the folding configuration.

The folding is enabled in the preference screen with the default values I did not touch it 
and the code is just a regular code written mostly by the eclipse wizard I just asked it to create a test class with a main method then I added a second blank function just to test the show only selected element

I tested it again with closing and opening the file and find something interesting the bug (meaning the disappearing on ending bracket bug) does not always happen

if I close the editor than reopen it when the show only selected element is on I get all the methods in the file and when I focus on one method using the outline view I get to see only that method with the ending bracket so in this case the bug does not occur

It is only when I open a file while the show only selected element is off then I click the icon in the toolbar to invoke it (and nothing changes) and then I click the method I want to focus on the outline view and I see only that method without the ending bracket so in this case the bug occur

So it seems that only if you invoke show only selected element on a file is already open and go to a method using the outline view the bug happen and the ending bracket disappear and this is a bug

should I report this bug ? and to who ?  

Edit : One good thing from all of this even with the bug folding now works when you show only selected element is enabled so the changes you merged even without the patch did make a great difference so Thank You for that

**Comment by danthe1st** [comment]
> It is only when I open a file while the show only selected element is off then I click the icon in the toolbar to invoke it (and nothing changes) and then I click the method I want to focus on the outline view and I see only that method without the ending bracket so in this case the bug occur

Yes, that is what I am saying the whole time. The unintended behavior (which I don't consider to be a problem) is that the feature works if you enable the preference while editors are open (in fact, restarting the IDE should result in folding being disabled as well until the other PR is merged). When I tried it, I didn't have the situation with the closing `}`  being hidden.

> should I report this bug ?

I don't think it makes much sense to create a new issue anywhere. If you have reliably steps to reproduce it, feel free to comment it in the aforementioned PR (I guess just commenting here is fine as well) but by just typing similar code, it didn't seem to do anything for me (maybe it's Windows-specific/only happens with CRLF, I didn't check that situation due to me using Linux - maybe I can check that at some point).

**Comment by danthe1st** [comment]
Yes, it's because of Windows line endings. I hope that I can take a look at that.

**Comment by totomomo18** [comment]
Thank you for your help I hope you fix it in the future I be waiting but you did make the folding work when show only selected is on so that is a great improvement and I be using that

**Comment by danthe1st** [comment]
@totomomo18 #3585 should fix that part.
